compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -839,40 +839,59 @@ object TypeOps:
       }
     }
 
+    object TraverseTp2 extends TypeTraverser:
+      val singletons = new mutable.ListBuffer[SingletonType]
+      val gadtSyms   = new mutable.ListBuffer[Symbol]
+      def traverse(tp: Type) =
+        val tpd = tp.dealias
+        if tpd ne tp then traverse(tpd)
+        else tp match
+          case tp: SingletonType if !singletons.contains(tp)=>
+            singletons += tp
+            traverseChildren(tp)
+          case tp: TypeRef if tp.symbol.isAbstractOrParamType =>
+            gadtSyms += tp.symbol
+            traverseChildren(tp)
+          case _ =>
+            traverseChildren(tp)
+    TraverseTp2.traverse(tp2)
+    val singletons = TraverseTp2.singletons.toList
+    val gadtSyms   = TraverseTp2.gadtSyms.toList
+
     // Prefix inference, replace `p.C.this.Child` with `X.Child` where `X <: p.C`
     // Note: we need to strip ThisType in `p` recursively.
     //
     // See tests/patmat/i3938.scala
     class InferPrefixMap extends TypeMap {
       var prefixTVar: Type | Null = null
       def apply(tp: Type): Type = tp match {
-        case ThisType(tref: TypeRef) if !tref.symbol.isStaticOwner =>
+        case ThisType(tref) if !tref.symbol.isStaticOwner =>
           val symbol = tref.symbol
-          if (symbol.is(Module))
+          val singletonIx = singletons.indexWhere(_.classSymbol == symbol)
+          if singletonIx >= 0 then
+            prefixTVar = singletons(singletonIx)
+            prefixTVar.uncheckedNN
```

Comment:
It seems to be possible that `tp2`  contains singletons in its type components that accidentally matches the prefix of `tp1`. But that would not matter in practice, users should not create such esoteric code.  Maybe add a comment mentioning it's intended for `tests/pos/i16785.scala`.

If it's possible to trigger the hang elsewhere in type checking, maybe also add that the fix is just a temporary workaround and should be removed once the deeper problem is resolved.

Reply:
Ah, yeah, I had that thought and forgot when I got to implementing it on tp2's side.  I'll try to see if I can avoid that problem and leave a test reference too.

> If it's possible to trigger the hang elsewhere in type checking, maybe also add that the fix is just a temporary workaround and should be removed once the deeper problem is resolved.

Root problem not withstanding I consider it better to be doing this in general: only add variables where needed.